### PR TITLE
Support 'go to definition' for self

### PR DESCRIPTION
**Pull request by Veykril**
Also reverts #6660, instead of showing the type it now works like it does for names by returning the declaration we are already on. This for example enables VSCode to show all references(#6665) when executing `go to definition` on the declaration.

### CODE REVIEW

File: crates/ide/src/goto_definition.rs [hunk 1]
```diff
@@ -987,31 +1009,31 @@ fn g() -> <() as Iterator<A = (), B<|> = u8>>::A {}
...
             r#"
 struct Foo {}
-     //^^^
 
 impl Foo {
-    fn bar(&self<|>) {}
-}
-"#,
-        );
+    fn bar(&self: &Foo) {
```
Comment (jonas-schievink) [inline crates/ide/src/goto_definition.rs:1018]: This isn't valid syntax, maybe this does the trick?

```suggestion
    fn bar(self: &Foo) {
```

**Comment by matklad** [comment]
I wonder what's the super correct behavior here.

* When "gotodef" on `self` reference, we should go to `self` parameter. 
* When "gotodef" on `self` parameter, we should find references
* "goto type def" on both should, well, go to the `struct` / `enum`

But I often also want to go the impl header instead.... Perhaps "goto type" should be changed to that?

**Comment by flodiebold** [comment]
Isn't there also "go to implementation"? We could reinterpret that :sweat_smile:

**Comment by matklad** [comment]
Lol yes, we can totally re-use goto impl here. Although I am a bit less sure about not re-using "go to type", beacause it doesn't have a clear impl imo.

I am thinking about this case:

```rust
struct Container<T>(T);

impl Container<String> {
    fn foo(&self) {}
}
```

Here, going to `Container<T>` loses the info about the fact that `T` is a `String`. The type of `self` is `Self`, which is a type-alias for the type declared in the impl's header. 

The example seems equivalent to 

```rust
type SelfType = Container<String>;
fn foo(this: &SelfType) {
}
```

to me.

Hm.... Now that I've spelled this out, I kinda lean towards changing "goto type", the alias thinking seems convincing to me (not that it is hard to convince oneself...)

**Comment by flodiebold** [comment]
That's a somewhat convincing argument, buut...
```rust
let me = self;
```
I'd expect "go to type" to work the same on `me` and `self` since they have the same type. Now this could actually work that way once we implement "lazy normalization" of type aliases; then go to type for `Self`-typed values could go to the impl, and it could also go to type aliases.

So my suggestion would be to make "go to impl" do this now, and "go to type" later when we can handle type aliases consistently there :smile:

**Pull request merged**